Make RecordFunction more robust for async use cases

ilia-cher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#34122 Make RecordFunction more robust for async use cases**

Summary:
Earlier work added support for async rpc cases when RecordFunction's
end callbacks might be called in a different thread; in addition some
extra care was needed to handle pointer to parent function;

This PR makes RecordFunction aware of potentially multiple threads in
use, as well as removes unused parent() call and restricts current()
RecordFunction to scope-based record functions (RECORD_FUNCTION macro)

Test Plan:
unit tests

Differential Revision: [D20297709](https://our.internmc.facebook.com/intern/diff/D20297709)

Commit:
Make RecordFunction more robust

Summary:
Earlier work added support for async rpc cases when RecordFunction's
end callbacks might be called in a different thread; in addition some
extra care was needed to handle pointer to parent function;

This PR makes RecordFunction aware of potentially multiple threads in
use, as well as removes unused parent() call and restricts current()
RecordFunction to scope-based record functions (RECORD_FUNCTION macro)

Test Plan:
unit tests

[ghstack-poisoned]

rohan-varma commented on torch/csrc/autograd/profiler.cpp:
```diff
@@ -162,9 +161,9 @@ void enableProfiler(ProfilerConfig config) {
-          // If we've overridden the thread_id on the RecordFunction, then find
-          //  the eventList that was created for the original thread_id. Then,
+        if (fn.getThreadId() != RecordFunction::getCurrentThreadId()) {
+          // If we've not in a thread that ran start callbacks, then find
```
Nit: If we're

rohan-varma commented on torch/csrc/autograd/record_function.h:
```diff
@@ -93,49 +93,49 @@ struct TORCH_API RecordFunction {
-  // The thread_id that this RecordFunction was created with. If 0, this means
-  // that it was not set with setThreadId() and this RecordFunction's callbacks
-  // cannot be invoked from a separate thread.
+  bool is_current_ = false;
```
nit: add a comment indicating that this flag is only true in the RECORD_FUNCTION macro for scope management?

rohan-varma commented on torch/csrc/autograd/record_function.cpp:
```diff
@@ -145,12 +148,28 @@ void runBeforeCallbacks(RecordFunction* rf, const std::string& funcName) {
+uint16_t RecordFunction::getCurrentThreadId() {
+  if (!current_thread_id_) {
+    // happens only once per thread
+    std::lock_guard<std::mutex> guard(next_thread_id_mutex_);
```
Are we sure that we can't deadlock in any case by different threads acquiring `getCurrentThreadId` and requiring the `all_event_list_map_mutex` in different orders?

ilia-cher replied:
we don't acquire locks under next_thread_id_mutex_ so deadlock is not possible

rohan-varma commented on torch/csrc/distributed/autograd/utils.cpp:
```diff
@@ -125,9 +125,6 @@ std::shared_ptr<FutureMessage> sendMessageWithAutograd(
   if (rf != nullptr) {
-    // save the local threadId so that end() callbacks can be correctly invoked
-    // from a different thread.
-    rf->setThreadId();
```
Nice, this will definitely make the API a lot cleaner for profiling async use cases

rohan-varma reviewed (approved):
Overall changes look good. Maybe someone with more familiarity around RecordFunction/profiler should take a look as well.

rohan-varma commented on torch/csrc/autograd/profiler.cpp:
```diff
@@ -162,9 +161,9 @@ void enableProfiler(ProfilerConfig config) {
-        if (fn.getThreadId() != 0) {
-          // If we've overridden the thread_id on the RecordFunction, then find
-          //  the eventList that was created for the original thread_id. Then,
+        if (fn.getThreadId() != RecordFunction::getCurrentThreadId()) {
```
Thinking about this more, maybe `fn.getThreadId()` is better named `getCreatingThreadId` or `getStartCallbacksThreadId` to indicate that it returns the thread id where this record function was created. I can do the renaming in a follow up PR.

ilia-cher replied:
sounds good, I'll update

Commit:
Update on "Make RecordFunction more robust for async use cases"


Summary:
Earlier work added support for async rpc cases when RecordFunction's
end callbacks might be called in a different thread; in addition some
extra care was needed to handle pointer to parent function;

This PR makes RecordFunction aware of potentially multiple threads in
use, as well as removes unused parent() call and restricts current()
RecordFunction to scope-based record functions (RECORD_FUNCTION macro)

Test Plan:
unit tests

[ghstack-poisoned]
